### ISSUE TITLE
Shutdown bug?

### ISSUE THREAD

**Issue by liviuchircu** [OpenSIPS/opensips#1347]
Occasionally, OpenSIPS may display the following behavior at shutdown (e.g., even on a simple `systemctl restart opensips`):

```
Apr 20 14:24:24 localhost /usr/sbin/opensips[6373]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6374]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6375]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6376]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6377]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6378]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6379]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6349]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 1(6346) [HTTPD INADDR_ANY:8080] terminated, still waiting for 33 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 2(6347) [MI FIFO] terminated, still waiting for 32 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 4(6349) [timer] terminated, still waiting for 31 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 21(6366) [SIP receiver hep_udp:127.0.0.1:6090 ] terminated, still waiting for 30 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 23(6368) [SIP receiver hep_udp:127.0.0.1:6090 ] terminated, still waiting for 29 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 22(6367) [SIP receiver hep_udp:127.0.0.1:6090 ] terminated, still waiting for 28 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 24(6369) [SIP receiver hep_udp:127.0.0.1:6090 ] terminated, still waiting for 27 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 25(6370) [TCP receiver] terminated, still waiting for 26 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 26(6371) [TCP receiver] terminated, still waiting for 25 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 27(6372) [TCP receiver] terminated, still waiting for 24 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 28(6373) [TCP receiver] terminated, still waiting for 23 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 29(6374) [TCP receiver] terminated, still waiting for 22 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 30(6375) [TCP receiver] terminated, still waiting for 21 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 32(6377) [TCP receiver] terminated, still waiting for 20 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 34(6379) [TCP main] terminated, still waiting for 19 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 31(6376) [TCP receiver] terminated, still waiting for 18 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 33(6378) [Timer handler] terminated, still waiting for 17 more
Apr 20 14:24:24 localhost /usr/sbin/opensips[6348]: INFO:core:sig_usr: signal 15 received
Apr 20 14:24:24 localhost /usr/sbin/opensips[6345]: INFO:core:shutdown_opensips: process 3(6348) [time_keeper] terminated, still waiting for 16 more
Apr 20 14:24:54 localhost systemd[1]: opensips.service: State 'stop-sigterm' timed out. Killing.
Apr 20 14:24:54 localhost systemd[1]: opensips.service: Main process exited, code=killed, status=9/KILL
Apr 20 14:24:54 localhost systemd[1]: Stopped OpenSIPS is a very fast and flexible SIP (RFC3261) server.
Apr 20 14:24:54 localhost systemd[1]: opensips.service: Unit entered failed state.
Apr 20 14:24:54 localhost systemd[1]: opensips.service: Failed with result 'signal'.
```

Notice how `systemd` is sending a `SIGKILL`, as a 30-second timeout. Just documenting here, so we don't forget about it. @rvlad-patrascu said he ran into this a couple of times as well.

Version: latest `2.3` from deb packages.

**Comment by bogdan-iancu** [comment]
According the logs, it looks like a 2.4 to me.

**Comment by liviuchircu** [comment]
My bad, it's a `2.4 nightly` deb.

**Comment by liviuchircu** [comment]
`ii  opensips                         2.4.0~20180419~818a74c-1`, so this is before the IPC fix!

**Comment by bogdan-iancu** [comment]
The shutdown procedure was not affected by the IPC bug . It looks like the other 16 remain processes are stuck in doing something else and fail to obey the RPC to terminate. The OpenSIPS attendant proc will give a 5 sec time window for all the procs to terminate, before using the heavy artillery and do a SIGTERM on them . It is interesting that I do not see any ""force termination for all processes" log after 5 secs from the start of the shutdown sequences.....
Can the behavior be reproduced ?

**Comment by liviuchircu** [comment]
Yes, I just reproduced it on latest nightly `2.4`. Just do:

```
systemctl restart opensips; ps aux | grep opensips
systemctl restart opensips; ps aux | grep opensips
systemctl restart opensips; ps aux | grep opensips
```

3-4 times is enough to trigger a bad shutdown.

**Comment by thesipguy** [comment]
It happens to me every time opensips 2.4 segfaults (approx every 2 minutes 👎 ) I need to kill all opensips processes before I can restart it.

**Comment by bogdan-iancu** [comment]
The issue identified in a syslog deadlock in the signal handlers, I will fix it asap.

**Comment by bogdan-iancu** [comment]
Fixed via 602089bbc6423055f04beb4e0cff2d36aa90dbdc, c00e9282f5df90abee8e34538e056817b21075bf and 4ee26e32554b9ebc149b432efbf3c6d7818840b5
